partiql-planner/src/main/kotlin/org/partiql/planner/typer/PlanTyper.kt

```diff
@@ -694,12 +699,50 @@ internal class PlanTyper(
             TODO("Type RexOpPivot")
         }
 
-        override fun visitRexOpCollToScalar(node: Rex.Op.CollToScalar, ctx: StaticType?): Rex {
-            TODO("Type RexOpCollToScalar")
+        override fun visitRexOpSubquery(node: Rex.Op.Subquery, ctx: StaticType?): Rex {
+            val select = visitRexOpSelect(node.select, ctx).op as Rex.Op.Select
+            val subquery = node.copy(select = select)
+            return when (node.coercion) {
+                Rex.Op.Subquery.Coercion.SCALAR -> visitRexOpSubqueryScalar(subquery, select.constructor.type)
+                Rex.Op.Subquery.Coercion.ROW -> visitRexOpSubqueryRow(subquery, select.constructor.type)
+            }
         }
 
-        override fun visitRexOpCollToScalarSubquery(node: Rex.Op.CollToScalar.Subquery, ctx: StaticType?): Rex {
-            TODO("Type RexOpCollToScalarSubquery")
+        /**
+         * Calculate output type of a row-value subquery.
+         */
+        private fun visitRexOpSubqueryRow(subquery: Rex.Op.Subquery, cons: StaticType): Rex {
+            if (cons !is StructType) {
+                return rexErr("Subquery with non-SQL SELECT cannot be coerced to a row-value expression. Found constructor type: $cons")
+            }
+            // Do a simple cardinality check for the moment.
+            // TODO we can only check cardinality if we know we are in a a comparison operator.
+            // val n = coercion.columns.size
+            // val m = cons.fields.size
+            // if (n != m) {
+            //     return rexErr("Cannot coercion subquery with $m attributes to a row-value-expression with $n attributes")
```

Comment:
I agree; I actually had this check in the RexConverter rather than the typer because I had access to the AST. I moved the to PlanTyper to perform type assertions on the constructor, but I'm beginning to think we may be limited to array expressions (not array literals).

Reply:
The cardinality check is just as easy for array expressions and literals, and we need the typing of the subquery for cases like

```
( 0, 0 ) = (SELECT * FROM points LIMIT 1); -- check if the first point is the origin
```